### PR TITLE
feat(chromium): large file uploads

### PR DESCRIPTION
**Pull request by yury-s**
Support uploading files larger than 50Mb (Chromium only, other browsers to follow). Some limitations: if any argument is >50Mb then all arguments should be file paths. In local case we just set the paths in the page, in the remote case we first copy the files to the remote machine.

#3768

### CODE REVIEW

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -1,11 +0,0 @@
-{
-    "name": "Playwright",
-    "image": "mcr.microsoft.com/playwright:next",
-    "postCreateCommand": "npm install && npm run build && apt-get update && apt-get install -y software-properties-common && curl -fsSL https://download.docker.com/linux/ubuntu/gpg | apt-key add - && add-apt-repository \"deb [arch=amd64] https://download.docker.com/linux/ubuntu $(lsb_release -cs) stable\" && apt-get install -y docker-ce-cli",
```
Comment (mxschmitt) [inline .devcontainer/devcontainer.json:4]: Please don't delete this file, we (lusha or me) use it with GitHub codespaces.

Comment (yury-s) [reply .devcontainer/devcontainer.json:4]: oops, this sneaked in after me fighting with misbehaving code formatting :)

File: packages/playwright-core/src/client/elementHandle.ts [hunk 2]
```diff
@@ -241,9 +248,37 @@ export function convertSelectOptionValues(values: string | api.ElementHandle | S
...
+  if (hasLargeFile && items.find(f => typeof f !== 'string'))
+    throw new Error('Mixing large files and FailePayload as argument is not supported, please pass all items as file paths.');
+
   const filePayloads: SetInputFilesFiles = await Promise.all(items.map(async item => {
+    if (hasLargeFile) {
+      assert(typeof item === 'string');
+      if (context._browser?._connection.isRemote()) {
+        const { writableStream: stream } = await context._channel.createTempFile({ name: path.basename(item) });
+        const writable = WritableStream.from(stream);
+        const pipelineAsync = promisify(pipeline);
```
Comment (mxschmitt) [inline packages/playwright-core/src/client/elementHandle.ts:270]: move this const up to line 32

File: packages/playwright-core/src/server/browserContext.ts [hunk 3]
```diff
@@ -274,6 +276,10 @@ export abstract class BrowserContext extends SdkObject {
     await Promise.all(Array.from(this._downloads).map(download => download.artifact.deleteOnContextClose()));
   }
 
+  private async _deleteAllTempDirs(): Promise<void> {
+    await Promise.all(this._tempDirs.map(async dir => await fs.promises.unlink(dir).catch(e => {})));
```
Comment (mxschmitt) [inline packages/playwright-core/src/server/browserContext.ts:280]: this sounds like we want to have it in the `packages/playwright-core/src/utils/processLauncher.ts>gracefullyCloseSet` as well.

File: tests/browsertype-connect.spec.ts [hunk 4]
```diff
@@ -572,3 +573,57 @@ test('should fulfill with global fetch result', async ({ browserType, startRemot
...
+  const remoteServer = await startRemoteServer();
+  const browser = await browserType.connect(remoteServer.wsEndpoint());
+  const context = await browser.newContext();
+  const page = await context.newPage();
+
+  await page.goto(server.PREFIX + '/input/fileupload.html');
+  const uploadFile = testInfo.outputPath('200MB.zip');
+  const str = 'A'.repeat(4 * 1024);
+  const stream = fs.createWriteStream(uploadFile);
+  for (let i = 0; i < 50 * 1024; i++) {
```
Comment (mxschmitt) [inline tests/browsertype-connect.spec.ts:589]: idea (might make not much sense): I personally would use a more complicated data payload, like counting numbers up and putting it in a string, by that we verify also the content in the data to spot off by one errors etc.

File: tests/page/page-set-input-files.spec.ts [hunk 5]
```diff
@@ -36,6 +36,53 @@ it('should upload the file', async ({ page, server, asset }) => {
...
+  });
+  const [file1] = await Promise.all([
+    serverFilePromise,
+    page.click('input[type=submit]')
+  ]);
+  expect(file1.originalFilename).toBe('200MB.zip');
+  expect(file1.size).toBe(200 * 1024 * 1024);
+  await Promise.all([uploadFile, file1.filepath].map(fs.promises.unlink));
+});
+
```
Comment (mxschmitt) [inline tests/page/page-set-input-files.spec.ts:85]: Seems like you added accidentally the `utils/linux-browser-dependencies/playwright-core.tar.gz` file. Feel free to add it to gitignore!

**Review by mxschmitt — APPROVED** [review]

File: packages/playwright-core/src/client/elementHandle.ts [hunk 6]
```diff
@@ -241,9 +248,37 @@ export function convertSelectOptionValues(values: string | api.ElementHandle | S
 }
 
 type SetInputFilesFiles = channels.ElementHandleSetInputFilesParams['files'];
-export async function convertInputFiles(files: string | FilePayload | string[] | FilePayload[]): Promise<SetInputFilesFiles> {
-  const items: (string | FilePayload)[] = Array.isArray(files) ? files : [ files ];
+export async function convertInputFiles(files: string | FilePayload | string[] | FilePayload[], context: BrowserContext): Promise<SetInputFilesFiles> {
+  const items: (string | FilePayload)[] = Array.isArray(files) ? files.slice() : [ files ];
+
+  const sizeLimit = 50 * 1024 * 1024;
```
Comment (dgozman) [inline packages/playwright-core/src/client/elementHandle.ts:254]: Let's log to pw:api "switching to large files mode".

File: packages/playwright-core/src/client/elementHandle.ts [hunk 7]
```diff
@@ -241,9 +248,37 @@ export function convertSelectOptionValues(values: string | api.ElementHandle | S
...
+  if (hasLargeBuffer)
+    throw new Error('Cannot set buffer larger than 50Mb, please write it to a file and pass its path instead.');
+
+  const stats = await Promise.all(items.filter(f => typeof f === 'string').map(item => fs.promises.stat(item as string)));
+  const hasLargeFile = !!stats.find(s => s.size > sizeLimit);
+  if (hasLargeFile && items.find(f => typeof f !== 'string'))
+    throw new Error('Mixing large files and FailePayload as argument is not supported, please pass all items as file paths.');
+
   const filePayloads: SetInputFilesFiles = await Promise.all(items.map(async item => {
+    if (hasLargeFile) {
```
Comment (dgozman) [inline packages/playwright-core/src/client/elementHandle.ts:265]: Let's do two separate protocol commands.

File: packages/playwright-core/src/dispatchers/frameDispatcher.ts [hunk 8]
```diff
@@ -200,7 +200,7 @@ export class FrameDispatcher extends Dispatcher<Frame, channels.FrameChannel> im
     return { values: await this._frame.selectOption(metadata, params.selector, elements, params.options || [], params) };
   }
 
-  async setInputFiles(params: channels.FrameSetInputFilesParams, metadata: CallMetadata): Promise<void> {
+  async setInputFiles(params: channels.FrameSetInputFilesParams, metadata: CallMetadata): Promise<channels.FrameSetInputFilesResult> {
     return await this._frame.setInputFiles(metadata, params.selector, params.files, params);
```
Comment (dgozman) [inline packages/playwright-core/src/dispatchers/frameDispatcher.ts:204]: You should turn all stream dispatchers into WriteableStream objects here.

**Review by dgozman — COMMENTED** [review]
Please remove utils/linux-browser-dependencies/playwright-core.tar.gz.

File: packages/playwright-core/src/client/elementHandle.ts (hunk 2 above)
Comment (yury-s) [reply packages/playwright-core/src/client/elementHandle.ts:270]: this is the only usage

File: tests/browsertype-connect.spec.ts (hunk 4 above)
Comment (yury-s) [reply tests/browsertype-connect.spec.ts:589]: I'd do that if I were to verify the content, so far it's just size comparison so should be good enough.

File: packages/playwright-core/src/server/browserContext.ts (hunk 3 above)
Comment (yury-s) [reply packages/playwright-core/src/server/browserContext.ts:280]: We careate the temp dirs in artifacts dir which is already managed and deleted by the launcher. This code juste does the cleanup earlier (we don't want to keep large file for longer than necessary and may even make the cleanup on navigation/page close in the future).

File: packages/playwright-core/src/client/elementHandle.ts [hunk 9]
```diff
@@ -241,8 +257,35 @@ export function convertSelectOptionValues(values: string | api.ElementHandle | S
...
+  const hasLargeBuffer = items.find(item => typeof item === 'object' && item.buffer && item.buffer.byteLength > sizeLimit);
+  if (hasLargeBuffer)
+    throw new Error('Cannot set buffer larger than 50Mb, please write it to a file and pass its path instead.');
+
+  const stats = await Promise.all(items.filter(f => typeof f === 'string').map(item => fs.promises.stat(item as string)));
+  const hasLargeFile = !!stats.find(s => s.size > sizeLimit);
+  if (hasLargeFile) {
+    if (context._browser?._connection.isRemote()) {
+      const streams: channels.WritableStreamChannel[] = await Promise.all(items.map(async item => {
+        assert(typeof item === 'string');
```
Comment (dgozman) [inline packages/playwright-core/src/client/elementHandle.ts:278]: isString

**Review by dgozman — APPROVED** [review]

**Pull request merged**